File: leetcode/1593. Split a String Into the Max Number of Unique Substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        res = 1
        def helper(start, onpath):
            if start == len(s):
                nonlocal res
                res = max(res, len(onpath))
                return
            
            for i in range(start, len(s)):
                curr = s[start:i+1]
                if curr in onpath:
                    continue
                onpath.add(curr)
                helper(i+1, onpath)
                onpath.remove(curr)
        helper(0, set())
        return res

File: leetcode/1593. Split a String Into the Max Number of Unique Substrings.kt
class Solution {
    fun maxUniqueSplit(s: String): Int {
        var res = 0
        fun splitMax(idx: Int, curr: MutableSet<String>) {
            if (idx == s.length) {
                res = max(res, curr.size)
                return
            }
            for (i in idx..s.length-1) {
                val sub = s.substring(idx, i+1)
                if (sub in curr) continue
                curr.add(sub)
                splitMax(i+1, curr)
                curr.remove(sub)
            }
        }
        splitMax(0, mutableSetOf<String>())
        return res
    }
}
